firedrake/function.py

```diff
@@ -356,6 +356,69 @@ def _interpolate_c_kernel(self, expression, to_pts, to_element, fs, coords):
                                             open_scope=False))
         return op2.Kernel(kernel_code, "expression_kernel")
 
+    def extract_subfunction(self, sub_domain):
+        """Extract a :class:`Function` equal to the current :class:`Function`
+        restricted to :arg:`sub_domain`. This currently only works for
+        the very limited case of extracting top and bottom functions
+        from CG fields on extruded domains. Future expansion to other
+        function spaces and subdomains will occur once function spaces
+        on subdomains have been implemented.
+
+        :param sub_domain: Either "top" or "bottom".
+        """
+
+        fs = self.function_space()
+
+        if not isinstance(fs.ufl_element(), ufl.OuterProductElement):
+            raise NotImplementedError(
+                "Can only extract subfunctions from extruded domains.")
+        if fs.ufl_element()._A.family() != "Lagrange" or \
+           fs.ufl_element()._B.family() != "Lagrange":
```

Comment:
This is no longer the right check for lagrange elements, since on quads the canonical name is "Q" not "Lagrange".  IOW, we need a better way of specifying these kind of checks.